### PR TITLE
internal/task: use asyncify on WASM

### PR DESCRIPTION
**Pull request by niaow**
This currently requires manually running the compiled wasm through `wasm-opt`.
Very little works so far.
Issue: #1101

### CODE REVIEW

**Comment by dgryski** [comment]
What work is remaining to get this PR finished?  I'd love to help out any way I can.

**Comment by niaow** [comment]
This is mainly blocked by #2035 I think

**Comment by dgryski** [comment]
While that bug is getting fixed, is there work that could be done on this issue even with having to run wasm-opt manually?

**Comment by aykevl** [comment]
I'm almost inclined to say that we should perhaps just get this finished (even without #2035). That means compiling for WebAssembly requires an extra binary - but I think that's only required when using the new Asincify based scheduler?

Thoughts?
I don't like adding an external dependency (it makes distribution a whole lot harder) but right now concurrency on WebAssembly is already somewhat broken and this PR might fix it.

**Comment by niaow** [comment]
Alright, I can modify this to just shell out to wasm-opt for now when using asyncify.

**Comment by niaow** [comment]
I made that change and attempted to debug some issues. It appears that after the initial launch suspend, resumed goroutines will not suspend again. I don't quite understand it yet.

**Comment by dgryski** [comment]
I'll take a look at this on Monday and see what I can come up with.

**Comment by niaow** [comment]
I have fixed what was probably the biggest bug. However, there are still some things left to be fixed apparently, such as weird WASI crashes and some sort of unicode error in the reflect test (. . . and i still havent added binaryen to CI).

**Comment by dgryski** [comment]
\o/. This works great on my little test `ping pong` program (included below)

```
package main

import (
	"time"
)

func pinger(pinger <-chan int, ponger chan<- int, who string) {
	for {
		<-pinger
		println(who)
		time.Sleep(time.Second)
		ponger <- 1
	}
}

func main() {
	ping := make(chan int)
	pong := make(chan int)

	go pinger(ping, pong, "ping")
	go pinger(pong, ping, "pong")

	ping <- 1

	select {}
}
```

However, I'm getting an unaligned pointer error with `gc=leaking` on `wasmtime` when trying to read the clock.

I managed to fix it with this patch.

```
 func ticks() timeUnit {
-       var nano uint64
-       clock_time_get(0, timePrecisionNanoseconds, &nano)
+       var b [16]byte
+       ptr := uintptr(unsafe.Pointer(&b[0]))
+       ptr += 8 + (^ptr & 0x7) + 1
+
+       clock_time_get(0, timePrecisionNanoseconds, (*uint64)(unsafe.Pointer(ptr)))
+       nano := *(*uint64)(unsafe.Pointer(ptr))
        return timeUnit(nano)
 }

```

It seems like a proper fix lies elsewhere, although I'm not sure where..

Trying to run the above ping-pong code with `gc=conservative` instead produces:

```
Caused by:
    0: failed to invoke command default
    1: wasm trap: uninitialized element
       wasm backtrace:
           0: 0x6c376 - <unknown>!<wasm function 266>
           1: 0x1489c - <unknown>!<wasm function 68>
           2: 0x2e336 - <unknown>!<wasm function 158>
           3: 0x2d77d - <unknown>!<wasm function 156>
           4: 0x2d636 - <unknown>!<wasm function 155>
```

I'm still looking into this..

**Comment by niaow** [comment]
> However, I'm getting an unaligned pointer error with `gc=leaking` on `wasmtime` when trying to read the clock.

I suspect this may be fixed by #2009. This PR is kinda old so I can rebase tommorow.

**Comment by dgryski** [comment]
Adding `-g` to the options provided to `wasm-opt` with the following patch:

```
diff --git a/builder/build.go b/builder/build.go
index df329bd7..7568e8ae 100644
--- a/builder/build.go
+++ b/builder/build.go
@@ -641,7 +641,7 @@ func Build(pkgName, outpath string, config *compileopts.Config, action func(Buil
                                default:
                                        return fmt.Errorf("unknown opt level: %q", config.Options.Opt)
                                }
-                               cmd := exec.Command("wasm-opt", "--asyncify",
+                               cmd := exec.Command("wasm-opt", "-g", "--asyncify",
                                        "--optimize-level", strconv.Itoa(optLevel),
                                        "--shrink-level", strconv.Itoa(shrinkLevel),
                                        executable, "--output", executable)
```

gives me a much better error:
```
Caused by:
    0: failed to invoke command default
    1: In func wasi_snapshot_preview1::args_sizes_get at write size: Pointer not aligned to 4: Region { start: 1066183, len: 4 }
       wasm backtrace:
           0: 0x320f - <unknown>!runtime.initAll
           1: 0x307f - runtime.run$1
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:19:10
           2: 0x301d - <goroutine wrapper>
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:18:2
           3: 0x3cb6 - <unknown>!tinygo_launch
           4:  0x538 - (*internal/task.Task).Resume
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/internal/task/task_asyncify.go:94:17
           5: 0x2edb - runtime.scheduler
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler.go:171:11
                     - runtime.run
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/scheduler_any.go:24:11
                     - _start
                           at /Users/dgryski/go/src/github.com/tinygo-org/tinygo/src/runtime/runtime_wasm_wasi.go:21:5

```

So indeed, something about asyncify is futzing with the alignment of the pointers and making things unhappy.

**Comment by dgryski** [comment]
I rebased my local copy against `dev` with no success.  Same pointer alignment issue.  I'll keep looking at this..

**Comment by dgryski** [comment]
Patching gc_leaking.go with
```
 func initHeap() {
-       // preinit() may have moved heapStart; reset heapptr
-       heapptr = heapStart
+       // preinit() may have moved heapStart; reset heapptr and make sure it's aligned
+       heapptr = align(heapStart)
 }
```

allows both gc=conservative and gc=leaking to run the pingpong example.

**Comment by niaow** [comment]
For some reason I thought chocolatey had binaryen. I guess I will need to switch over the install then.

**Comment by dgryski** [comment]
Something about the asyncify code is causing https://github.com/tinygo-org/tinygo/issues/2172 to fail

File: azure-pipelines.yml
Comment (aykevl) [inline azure-pipelines.yml]: There is no reason to do this when not building LLVM.

File: compiler/compiler_test.go
Comment (aykevl) [inline compiler/compiler_test.go]: I think this should match the target JSON for consistency.

```suggestion
				DefaultStackSize:   config.DefaultStackSize(),
```

File: src/internal/task/task_asyncify.go
Comment (aykevl) [inline src/internal/task/task_asyncify.go]: Normally, I would expect an aligned stack to be passed in. For example, a stack size of 2048 is very much aligned.
I never really thought about aligning it, but yeah, technically it might need to be aligned.

I don't think this needs to be done at runtime (although there is a good chance it's optimized away in this case): it's only ever specified in code (in compileopts/target.go for example) or in target JSON files. The only way to easily change it is by modifying the JSON file directly, for example. If we're going to align it, I think it would be better to do it at compile time instead of at runtime, for example here:

https://github.com/tinygo-org/tinygo/blob/157382600535c95bdbf5d6e8c4b9e5ca71749018/compiler/goroutine.go#L125

This can be done by aligning to 16 bytes, I believe that's nowadays a pretty standard alignment (even on riscv32 and i386, for example).

File: src/internal/task/task_asyncify.go
Comment (aykevl) [inline src/internal/task/task_asyncify.go:94]: So I guess once this call returns, we're back after rewinding?

File: targets/wasi.json
Comment (aykevl) [inline targets/wasi.json]: This GC was already the default, so this is not technically needed. But many other targets also do it so :shrug:

File: testdata/channel.go
Comment (aykevl) [inline testdata/channel.go]: Why did you increase this to a second? That's some valuable time spent in a test. If you want to make it longer than a microsecond, please consider making it multiple microseconds or even a whole millisecond.

**Review by aykevl — COMMENTED** [review]
This looks pretty clean. I would like to take a more closer look at some parts, but so far I only have some relatively minor comments.

File: testdata/channel.go
Comment (niaow) [reply testdata/channel.go]: This was a temporary change I made when debugging a timer queue crash. I will revert this when cleaning up.

File: src/internal/task/task_asyncify.go
Comment (niaow) [reply src/internal/task/task_asyncify.go:94]: Yes.

**Comment by aykevl** [comment]
> For some reason I thought chocolatey had binaryen. I guess I will need to switch over the install then.

`scoop` has it, not `choco`.

**Comment by niaow** [comment]
I have created a fix for the WASM scheduler recursion failure: #2178

**Comment by dgryski** [comment]
I think this branch still has an odd failure.  Trying to build the following program with
`tinygo build -scheduler=asyncify -gc=leaking -target=wasi -o m.wasm main.go`


```
package main

import (
	"context"
)

func main() {
	ctx := context.Background()

	select {
	case <-ctx.Done():
	}
}
```

produces the following build failure:
```
Call parameter type does not match function signature!
%runtime.channel.56* null
 %runtime.channel.80*  %1 = call i1 bitcast (i1 (%runtime.channel*, i8*, %runtime.channelBlockedList*, i8*, i8*)* @runtime.chanRecv to i1 (%runtime.channel.80*, i8*, %runtime.channelBlockedList.79*, i8*, i8*)*)(%runtime.channel.56* null, i8* null, %runtime.channelBlockedList.79* %chan.blockedList, i8* undef, i8* null), !dbg !1627
```

**Comment by niaow** [comment]
I have been poking around, and it currently appears that the interface lowering pass is the first pass introducing type errors when building that code.

**Comment by niaow** [comment]
I identified the issue and seperated it into #2197

**Comment by niaow** [comment]
Hmm, it looks like LLVM 10 may have issues with the not-yet-resolved symbols for asyncify. I am not really sure how we want to handle this. I guess we could disable this on LLVM 10? I am not sure of a better option.

**Comment by dgryski** [comment]
Given that llvm 13 is out now, how many versions behind is tinygo committed to supporting?

**Comment by aykevl** [comment]
> Hmm, it looks like LLVM 10 may have issues with the not-yet-resolved symbols for asyncify. I am not really sure how we want to handle this. I guess we could disable this on LLVM 10? I am not sure of a better option.

Don't let that be a blocker. If LLVM 10 is an issue, we'll drop LLVM 10 support. That's a really old version by now. (Usually we support the last two LLVM versions or so, to make things easier for distributions). Even more so because this PR is a blocker to add support for newer LLVM versions.

**Comment by aykevl** [comment]
I've created #2216 to drop support for LLVM 10 so that this PR can continue without being burdened by that LLVM version.

**Comment by niaow** [comment]
Alright, sounds good. I will take a look at it tomorrow.
The main other issues I still need to deal with right now are:
1. Fix Windows CI
2. Figure out why `test-llvm11-go116` is timing out

File: azure-pipelines.yml
Comment (aykevl) [inline azure-pipelines.yml]: Apparently wget isn't included on this Windows version: https://github.com/actions/virtual-environments/blob/main/images/win/Windows2016-Readme.md
You may have better luck with curl, or by installing emscripten using Scoop instead: ScoopInstaller/Main#2740 (installing Scoop should be pretty fast).

**Comment by deadprogram** [comment]
Now that #2216 has been merged into `dev` branch, this PR should be able to continue. :smile_cat:

**Comment by niaow** [comment]
So it appears that we may need to require binaryen version 102 instead of 101. This makes it a bit more annoying to install, since 102 does not have pre-built binaries.

**Comment by aykevl** [comment]
> So it appears that we may need to require binaryen version 102 instead of 101.

Are you referring to this failure?

```
build/tinygo build             -o test.wasm -tags=hifive1b             examples/blinky1
error: wasm-opt failed: signal: segmentation fault (core dumped)
```

**Comment by niaow** [comment]
> Are you referring to this failure?

Yeah, I think that is fixed in binaryen 102. I am not 100% sure, although it seems to match up with a bug report in 101, and I have not been able to replicate any similar fault on 102.

**Comment by deadprogram** [comment]
Any action on this PR? Quite a few people are eagerly awaiting it (see what I did there?).

**Comment by niaow** [comment]
> Any action on this PR? Quite a few people are eagerly awaiting it (see what I did there?).

Sorry I haven't gotten a chance to work on it this week.
Tomorrow I should hopefully be able to get back to it.
I am going to try to get all of the targets to build binaryen 102 from source and bundle it in the release tarball.

**Comment by niaow** [comment]
Getting this to work on stretch is seemingly more difficult than I anticipated, due to the requirement for a newer cmake version.

**Comment by aykevl** [comment]
Did you verify that the bug is indeed there in version 101 and fixed in 102? It could be something unrelated to the version.

**Comment by aykevl** [comment]
With version 101:

```
$ $WASMOPT --version
wasm-opt version 101 (version_101)

$ tinygo build -o test.wasm -tags=hifive1b -size=short examples/blinky1
   code    data     bss |   flash     ram
  29870     771  130301 |   30641  131072
```

I'm compiling version 102 right now.

**Comment by niaow** [comment]
> Did you verify that the bug is indeed there in version 101 and fixed in 102? It could be something unrelated to the version.

It is a non-deterministic bug, and does not replicate on my system. It doesn't seem to replicate on 102 so far, so it may be right it may be wrong but cant tell for sure.

**Comment by aykevl** [comment]
Hmm, perhaps with was a flaky test then?
If it can't be reproduced, I'd say don't worry about bundling wasm-opt in all TinyGo releases.

**Comment by niaow** [comment]
Not a flaky test. This is a legitimate bug, and reproduces randomly in CI (generally at least one of the builds fails, but rarely more than that) with 101.

**Comment by niaow** [comment]
I think(?) I have fixed all of the build errors at this point. I just need to wait for all of the CI run (and Windows may take a while).

**Comment by niaow** [comment]
I had removed the `rm -rf llvm-source` without realizing why it was done. I updated the comments in the CI config to try to explain it better, because it was surprising behavior.

**Comment by deadprogram** [comment]
All tests are passing now!

**Comment by niaow** [comment]
@deadprogram windows smoke tests are still running: https://github.com/niaow/tinygo/runs/4132134414?check_suite_focus=true

**Comment by niaow** [comment]
Windows smoke tests failed.

**Comment by niaow** [comment]
It looks like I messed up the autodetection on Windows (I need to add a `.exe` suffix).

**Comment by niaow** [comment]
The commit I just pushed should fix it, although it will take 2 hours to find out. I think this is ready for review to start now.

File: .circleci/config.yml
Comment (aykevl) [inline .circleci/config.yml:30]: Can you do this installation only when these tools are needed in the build-binaryen step? That speeds up the CI in the common case.

File: .circleci/config.yml
Comment (aykevl) [inline .circleci/config.yml:190]: Same here.

File: .circleci/config.yml
Comment (aykevl) [inline .circleci/config.yml:255]: And here.

File: .circleci/config.yml
Comment (aykevl) [inline .circleci/config.yml:330]: And here.

File: .circleci/config.yml
Comment (aykevl) [inline .circleci/config.yml]: Suggestion: this might be clearer when the name is `build-binaryen-linux`. Because the cache is only for Linux.
Similarly, I think it's better to rename the cache key to something like `binaryen-linux-v1`.

File: .gitignore
Comment (aykevl) [inline .gitignore:1]: Oh, I see you've moved the .gitignore line into a separate file in the build directory.
I guess that's fine, just wondering what the reason is?

File: Makefile
Comment (aykevl) [inline Makefile]:
```suggestion
	cd lib/binaryen && cmake -G Ninja . -DBUILD_STATIC_LIB=ON && ninja
```

Not sure whether this works as-is in a Makefile but generally not continuing on errors is a good thing.

**Review by aykevl — COMMENTED** [review]
I've left a few comments about how the CI can be sped up a bit, but in general this looks very good.

I have one question though: what happens if you export a function and do a blocking call in there? JavaScript can't block, so it must return _something_ instead. This could be surprising to users, especially if this behavior is not part of the function signature.

**Comment by aykevl** [comment]
> The commit I just pushed should fix it, although it will take 2 hours to find out. I think this is ready for review to start now.

Yeah, caching on GitHub Actions is a bit weird. It should work, but for some reason there are a lot of cache misses.

File: .gitignore
Comment (niaow) [reply .gitignore:1]: I needed to keep the `build` directory, and for some reason was struggling to get it to track as-is (negated patterns seemed to be behaving weirdly). This is just how I normally do that.

File: .circleci/config.yml
Comment (niaow) [reply .circleci/config.yml:30]: I did this like this for 2 reasons:
1. We need to install these differently on buster and stretch.
2. We need these for both LLVM and Binaryen, and avoiding a double install seemed annoying.

I guess I could potentially restore both caches before installing, although it seems to make this build config a lot more complicated.

Comment (niaow) [reply .circleci/config.yml:30]: This appears to only make CI 1 second slower. I don't think the extra complexity is worthwhile.

**Comment by niaow** [comment]
> I have one question though: what happens if you export a function and do a blocking call in there? JavaScript can't block, so it must return something instead. This could be surprising to users, especially if this behavior is not part of the function signature.

Assuming that you use `js.FuncOf`. . .  the current behavior (unchanged by this PR) is that stuff will break. Go defines this to deadlock, but our implementation is still non-conformant (I hope to fix this in a future PR).

As for raw exported functions, they will panic in `task.Pause` if there are no running goroutines. If there are running goroutines (Go calls into JS calls into Go which then blocks), this will break and bad things will happen. This particular issue is a lot more difficult to solve, and will require us to provide our own implementation of `syscall/js` with some changes.

File: .circleci/config.yml
Comment (aykevl) [reply .circleci/config.yml:30]: > This appears to only make CI 1 second slower. I don't think the extra complexity is worthwhile.

How did you measure this? It seems rather unlikely that installing software is that fast.
(If it is really just 1 second, then installing it unconditionally would be fine).

Comment (niaow) [reply .circleci/config.yml:30]: I looked at 2 a build on the `dev` branch and a build from this PR. The "install apt dependencies steps" took 22 seconds on this PR and 23 seconds on the dev branch. Looking through more CI runs on `dev` now I see a lot more variation, but the difference in download size is 180 MB to 186 MB which gives me a seemingly appropriate estimate of 3%. These packages are relatively small compared to the other packages which we install.

**Comment by aykevl** [comment]
> As for raw exported functions, they will panic in `task.Pause` if there are no running goroutines. If there are running goroutines (Go calls into JS calls into Go which then blocks), this will break and bad things will happen.

Right. That was what I was wondering about. I suspect many people today call exported functions from TinyGo, as described [here](https://tinygo.org/docs/guides/webassembly/). Maybe `-scheduler=none` is the solution in that case, but maybe that's a bit too drastic. In any case, we have to provide some answer to the question of how to call into Go functions from JavaScript.
(This is why I prefer to keep `-scheduler=coroutines` around for one release if possible).

**Comment by niaow** [comment]
Something seems really wrong with the Windows build, so that might take a bit to debug. https://github.com/niaow/tinygo/runs/4132565997?check_suite_focus=true#step:17:72

(This means that binaryen crashed when just run with `--version`)

File: .circleci/config.yml
Comment (aykevl) [reply .circleci/config.yml:30]: Ok. Well maybe it is fine then. We can revisit this later, especially when we move away from Debian stretch.

**Comment by aykevl** [comment]
@niaow you might want to rebase on top of the dev branch, #2245 should avoid rebuilding LLVM on each push.

**Comment by niaow** [comment]
It looks like the version of GCC used by Debian Stretch is too old to build `binaryen` correctly, and that the cache was hiding this fact.

**Comment by niaow** [comment]
I managed to work around that by using the copy of clang that we build.
Now the only remaining issue is Windows.

**Comment by niaow** [comment]
If my debugging attempts are correct, it looks like binaryen is still trying to dynamically link against `libstdc++` and `libgcc` from mingw, and failing (error code `0xc0000139`).
```
[niaow@finch bin]$ objdump -p wasm-opt.exe | grep "DLL Name:"
	DLL Name: libgcc_s_seh-1.dll
	DLL Name: KERNEL32.dll
	DLL Name: msvcrt.dll
	DLL Name: libwinpthread-1.dll
	DLL Name: USER32.dll
	DLL Name: libstdc++-6.dll
```

**Comment by aykevl** [comment]
For the TinyGo build I use `-static -static-libgcc -static-libstdc++`, I think you might need those for binaryen too.

**Comment by niaow** [comment]
Yeah I just realized that. Adding now.

**Comment by aykevl** [comment]
I can try rebooting to Windows to test some build options, would that help?

**Comment by niaow** [comment]
I suspect that I just fixed it. If not, I finally have a Windows VM setup so I can test there.

**Comment by niaow** [comment]
All tests are now passing.

**Comment by niaow** [comment]
I did a quick squash/cleanup, so I think this is ready for a final review.

**Comment by deadprogram** [comment]
Do we need to modify the dev Dockerfile if you want to use this scheduler? https://github.com/tinygo-org/tinygo/blob/dev/Dockerfile#L23-L28

File: Dockerfile
Comment (deadprogram) [inline Dockerfile:34]: Oops, just noticed this. Never mind my other question.

File: Makefile
Comment (deadprogram) [inline Makefile]: I'm almost embarrassed to point out something so trivial as an extra tab here :joy_cat:

Comment (niaow) [reply Makefile]: Huh. I guess this is what happens when using 4-wide tabs in editor :facepalm:

File: Dockerfile
Comment (niaow) [reply Dockerfile:34]: I just double checked and this doesn't actually work anymore, so I do actually need to fix it.

**Review by deadprogram — APPROVED** [review]
All tests are now passing, and seems good to me.

**Comment by deadprogram** [comment]
@aykevl any last comments before merge?

**Comment by gedw99** [comment]
Is this the fix to get WASM compilation working again ?

**Comment by deadprogram** [comment]
> Is this the fix to get WASM compilation working again ?

WASM compilation is working afaik. This PR is to add asyncify as an option for task scheduling for WASM programs.

**Comment by gedw99** [comment]
> > Is this the fix to get WASM compilation working again ?
> 
> WASM compilation is working afaik. This PR is to add asyncify as an option for task scheduling for WASM programs.

ok thank for the response. 

https://github.com/tinygo-org/tinygo/issues/1972 is blocking i believe. I think is for gio when building to Web ( wasm ).

**Review by aykevl — COMMENTED** [review]
This looks good to me.
I can't really comment on the Dockerfile changes, is the issue there fixed now? I don't know a whole lot about Docker.

File: goenv/goenv.go
Comment (aykevl) [inline goenv/goenv.go:91]: [this comment was intended to be part of the review but I forgot to click the "start review" button]

I think this function can be simplified. In short: the bundled wasm-opt binaries don't need to be checked for a version.

I would suggest the following logic:

```
if $TINYGOROOT/bin/wasm-opt[.exe] exists:
    return this path
if $TINYGOROOT/build/wasm-opt[.exe] exists:
    return this path
look for wasm-opt using LookPath
if not found:
    print or return error
check for version
if version too old:
    return error
return this path
```

This is not a blocker to me.

Comment (niaow) [reply goenv/goenv.go:91]: I think we should still be doing a version check regardless for now, until we can actually link it in properly.

**Comment by niaow** [comment]
The Dockerfile issues have been fixed.

File: goenv/goenv.go
Comment (niaow) [reply goenv/goenv.go:91]: This has already caught one issue, so it is still helpful.

**Comment by deadprogram** [comment]
OK merge time! Great work on this @niaow thank you. Also thanks to everyone who helped review/test.

**Review by deadprogram — APPROVED** [review]

**Pull request merged**
